src/handler/TokenHandler.ts

```diff
@@ -80,18 +91,50 @@ class TokenHandler {
     ) {
       try {
         var resp = await this.getTokensFromRefreshToken();
+        idToken = resp.id_token;
       } catch (err) {
         // No refresh token in cookie or some other error
         return null;
       }
-      idToken = resp.id_token;
     }
     this.idToken = idToken;
     return new CotterIDToken(idToken);
   }
 
-  // Returns access token
+  // Refresh tokens only when needed.
   async getTokensFromRefreshToken(): Promise<OAuthToken> {
+    return new Promise<OAuthToken>((resolve, reject) => {
+      const checkTokenFetchProcess = () => {
+        if (this.tokenFetchingState === TOKEN_FETCHING_STATES.initial) {
+          this.getTokensFromRefreshTokenAPI();
+        } else if (
+          this.tokenFetchingState === TOKEN_FETCHING_STATES.ready &&
+          this.fetchTokenResp
+        ) {
+          resolve(this.fetchTokenResp);
+          return;
+        } else if (
+          this.tokenFetchingState === TOKEN_FETCHING_STATES.errorRetry
+        ) {
+          this.tokenFetchingState = TOKEN_FETCHING_STATES.initial;
+          reject();
+          return;
+        } else if (
+          this.tokenFetchingState === TOKEN_FETCHING_STATES.errorFatal
+        ) {
+          reject();
+          return;
+        }
```

Comment:
can you put this conditional logic in a switch statement for better readability?